### PR TITLE
Define basic Linux SocketCAN constants and types

### PR DESCRIPTION
**Pull request by djarek**
Add definitions from `linux/can.h`, which is a "base" header for remainder of SocketCAN functionality.

CAN bus (ISO-11898) is a communication standard used in automotive, automation and industrial solutions. Linux provides a socket-like interface to access raw CAN and protocols based on CAN, such as ISO-TP(ISO-15765) or SAE-J1939.

### CODE REVIEW

File: src/unix/linux_like/linux/mod.rs [hunk 1]
```diff
@@ -504,6 +508,43 @@ s! {
         pub ee_info: u32,
         pub ee_data: u32,
     }
+
+    // linux/can.h
+    #[repr(align(8))]
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:513]: Should be behind `#[cfg(libc_align)]`.

File: src/unix/linux_like/linux/mod.rs [hunk 2]
```diff
@@ -504,6 +508,43 @@ s! {
...
+    pub struct can_frame {
+        pub can_id: canid_t,
+        pub can_dlc: u8,
+        pub __pad: u8,
+        pub __res0: u8,
+        pub __res1: u8,
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:519]: They don't have to be public?
```suggestion
        __pad: u8,
        __res0: u8,
        __res1: u8,
```

File: src/unix/linux_like/linux/mod.rs [hunk 3]
```diff
@@ -504,6 +508,43 @@ s! {
...
+        pub data: [u8; CAN_MAX_DLEN],
+    }
+
+    #[repr(align(8))]
+    pub struct canfd_frame {
+        pub can_id: canid_t,
+        pub len: u8,
+        pub flags: u8,
+        pub __res0: u8,
+        pub __res1: u8,
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:529]:
```suggestion
        __res0: u8,
        __res1: u8,
```

**Review by JohnTitor — COMMENTED** [review]
Overall LGTM, left some comments.

File: src/unix/linux_like/linux/mod.rs [hunk 4]
```diff
@@ -504,6 +508,45 @@ s! {
         pub ee_info: u32,
         pub ee_data: u32,
     }
+
+    // linux/can.h
+    #[cfg(libc_align)]
```
Comment (JohnTitor) [inline src/unix/linux_like/linux/mod.rs:513]: We have `align.rs` for that.

**Comment by JohnTitor** [comment]
Also, it'd be great if you could squash commits into one.

**Comment by djarek** [comment]
Squashed the commits.

**Comment by JohnTitor** [comment]
Thanks!
@bors r+

**Comment by JohnTitor** [comment]
CI failed:
```
   cargo:warning=/checkout/target/mips64el-unknown-linux-gnuabi64/debug/build/libc-test-7d5e3789cf3be7ba/out/main.c:31010:59: error: 'CAN_J1939' undeclared here (not in a function)
  cargo:warning=31010 |             static const int __test_const_CAN_J1939_val = CAN_J1939;
  cargo:warning=      |                                                           ^~~~~~~~~
  cargo:warning=cc1: error: unrecognized command line option '-Wno-unknown-warning-option' [-Werror]
  cargo:warning=cc1: all warnings being treated as errors
  exit code: 1
```

That target's CI stuff is old (since `CAN_J1939` is a new constant) or doesn't have it actually?

**Comment by djarek** [comment]
Looks like J1939 was added in Linux 5.4, not sure what version of kernel headers that target has.

**Comment by JohnTitor** [comment]
Let's skip the tests for now, we could tweak here:
https://github.com/rust-lang/libc/blob/9d6d0186e92304e57aafa5e5724ce37c269cee36/libc-test/build.rs#L2595

Add this with FIXME comment:

```rust
"CAN_J1939" if mips => true,
```

**Comment by JohnTitor** [comment]
Hm, it still failed, let me check with a quick tweak...

File: libc-test/build.rs [hunk 5]
```diff
@@ -2587,7 +2591,9 @@ fn test_linux(target: &str) {
             // can be an anonymous struct, so an extra struct,
             // which is absent in musl, has to be defined.
             "__exit_status" if musl => true,
-
+            // FIXME: J1939 requires kernel header version 5.4 or higher,
+            // the MIPS CI target has a lower version
+            "CAN_J1939" if mips => true,
```
Comment (JohnTitor) [inline libc-test/build.rs:2596]:
```suggestion
            "CAN_J1939" => true,
```

**Comment by JohnTitor** [comment]
Okay, figured out, `CAN_J1939` wasn't placed on `skip_const` but `skip_struct` therefore it wasn't skipped actually. Now it should pass CI.

@bors r+

**Comment by JohnTitor** [comment]
Skip more items as musl kernel header on CI is old.
@bors r+

**Pull request merged**